File: Leetcode - Java/99_tree_recoverTree.java
// https://leetcode.com/problems/recover-binary-search-tree/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode firstElem = null;
    TreeNode secondElem = null;
    TreeNode prevElem = new TreeNode(Integer.MIN_VALUE);
    
    public void recoverTree(TreeNode root) {
        traverse(root);
        int temp = firstElem.val;
        firstElem.val = secondElem.val;
        secondElem.val = temp;
    }
    
    private void traverse(TreeNode root) {
        if (root == null) return;
        traverse(root.left);
        if (firstElem == null && prevElem.val > root.val) {
            firstElem = prevElem;
        }
        if (firstElem != null && prevElem.val > root.val) {
            secondElem = root;
        }
        prevElem = root;
        traverse(root.right);
    }
}

File: Leetcode - Python/99_tree_recoverTree.py
# https://leetcode.com/problems/recover-binary-search-tree/

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def __init__(self):
        self.firstElem = self.secondElem = None
        self.prevElem = TreeNode(-2147483648)

    def recoverTree(self, root: TreeNode) -> None:
        self.traverse(root)
        self.firstElem.val, self.secondElem.val = self.secondElem.val, self.firstElem.val

    def traverse(self, root: TreeNode) -> None:
        if root is None:
            return

        self.traverse(root.left)
        if self.firstElem is None and self.prevElem.val > root.val:
            self.firstElem = self.prevElem
        if self.firstElem is not None and self.prevElem.val > root.val:
            self.secondElem = root
        self.prevElem = root
        self.traverse(root.right)
